### ISSUE TITLE
Support Pairwise Loss

### ISSUE THREAD

**Issue by rejasupotaro** [rejasupotaro/master-thesis#7]
## Overview

I'm not sure which loss I use at this moment but enabling the system to support multiple losses would make running experiments easier.

In this task, I implement a pairwise loss class and enable the system to switch the loss by configuration.

## Data Format

- https://github.com/keras-team/keras/issues/4861
- https://github.com/NTMC-Community/MatchZoo/blob/master/matchzoo/data_generator/data_generator.py

### Cross-Entropy Loss

```python
class PairwiseLogisticLoss(_PairwiseLoss):
  """Implements pairwise logistic loss."""

  def _pairwise_loss(self, pairwise_logits):
    """See `_PairwiseLoss`."""
    # The following is the same as log(1 + exp(-pairwise_logits)).
    return tf.nn.relu(-pairwise_logits) + tf.math.log1p(
        tf.exp(-tf.abs(pairwise_logits)))
```

- https://www.kaggle.com/c/porto-seguro-safe-driver-prediction/discussion/43842
- https://github.com/tensorflow/ranking/blob/6c55d0bc87c761448f1937f46998039c8dbfe280/tensorflow_ranking/python/losses_impl.py#L579
- https://github.com/NTMC-Community/MatchZoo/blob/master/matchzoo/losses/rank_cross_entropy_loss.py

### Hinge Loss

```python
class PairwiseHingeLoss(_PairwiseLoss):
  """Implements pairwise hinge loss."""

  def _pairwise_loss(self, pairwise_logits):
    """See `_PairwiseLoss`."""
    return tf.nn.relu(1 - pairwise_logits)
```

- https://github.com/tensorflow/ranking/blob/6c55d0bc87c761448f1937f46998039c8dbfe280/tensorflow_ranking/python/losses_impl.py#L579
- https://keras.io/api/losses/hinge_losses/

## Implementation

RankingLoss > PointwiseLoss, PairwiseLoss, ListwiseLoss

## Links

- [Understanding Ranking Loss, Contrastive Loss, Margin Loss, Triplet Loss, Hinge Loss and all those confusing names](https://gombru.github.io/2019/04/03/ranking_loss/)
- https://github.com/rejasupotaro/snrm-tf2/blob/master/snrm.py
- https://keras.io/examples/structured_data/collaborative_filtering_movielens/
- [[1811.04415] Learning Groupwise Multivariate Scoring Functions Using Deep Neural Networks](https://arxiv.org/abs/1811.04415)
